drivers/tee/tee_core.c

```diff
@@ -111,6 +121,209 @@ static int tee_release(struct inode *inode, struct file *filp)
+		   size_t size)
+{
+	unsigned char hash[SHA1_DIGEST_SIZE];
+	struct crypto_shash *shash = NULL;
+	struct shash_desc *desc = NULL;
+	int rc;
+
+	shash = crypto_alloc_shash("sha1", 0, 0);
+	if (IS_ERR(shash)) {
+		rc = PTR_ERR(shash);
+		pr_err("shash(sha1) allocation failed\n");
+		return rc;
+	}
+
+	desc = kzalloc(sizeof(*desc) + crypto_shash_descsize(shash),
+		       GFP_KERNEL);
+	if (IS_ERR(desc)) {
+		rc = PTR_ERR(desc);
+		goto out;
+	}
+
+	desc->tfm = shash;
+
+	rc = crypto_shash_init(desc);
+	if (unlikely(rc < 0))
+		goto out2;
+
+	rc = crypto_shash_update(desc, (const u8 *)ns, sizeof(*ns));
+	if (unlikely(rc < 0))
+		goto out2;
+
+	rc = crypto_shash_update(desc, (const u8 *)name, size);
+	if (unlikely(rc < 0))
+		goto out2;
+
+	rc = crypto_shash_final(desc, hash);
+	if (unlikely(rc < 0))
+		goto out2;
+
+	// time_low
+	uuid->b[0] = hash[0];
+	uuid->b[1] = hash[1];
+	uuid->b[2] = hash[2];
+	uuid->b[3] = hash[3];
+
+	// time_mid
+	uuid->b[4] = hash[4];
+	uuid->b[5] = hash[5];
+
+	// time_hi_and_version (v5)
+	uuid->b[6] = (hash[6] & 0x0F) | 0x50;
+	uuid->b[7] = hash[7];
+
+	// clock_seq_hi_and_reserved (variant)
+	uuid->b[8] = (hash[8] & 0x3F) | 0x80;
+
+	// clock_seq_low
+	uuid->b[9] = hash[9];
+
+	// node
+	uuid->b[10] = hash[10];
+	uuid->b[11] = hash[11];
+	uuid->b[12] = hash[12];
+	uuid->b[13] = hash[13];
+	uuid->b[14] = hash[14];
+	uuid->b[15] = hash[15];
+
+out2:
+	kfree(desc);
+
+out:
+	crypto_free_shash(shash);
+	return rc;
+}
+
+int tee_session_calc_client_uuid(uuid_t *uuid, u32 connection_method,
+				 const u8 connection_data[TEE_IOCTL_UUID_LEN])
+{
+	const char *application_id = NULL;
+	gid_t ns_grp = (gid_t)-1;
+	kgid_t grp = INVALID_GID;
+	char *name = NULL;
+	int rc;
+
+	if (connection_method == TEE_IOCTL_LOGIN_PUBLIC) {
+		/* Nil UUID to be passed to TEE environment */
+		uuid_copy(uuid, &uuid_null);
+		return 0;
+	}
+
+	/*
+	 * In Linux environment client UUID is based on UUIDv5.
+	 *
+	 * Determine client UUID with following semantics for 'name':
+	 *
+	 * For TEEC_LOGIN_USER:
+	 * user=<uid>
+	 *
+	 * For TEEC_LOGIN_GROUP:
+	 * group=<gid>
+	 *
+	 * For TEEC_LOGIN_APPLICATION:
+	 * app=<application id>
+	 *
+	 * For TEEC_LOGIN_USER_APPLICATION:
+	 * user=<uid>:app=<application id>
+	 *
+	 * For TEEC_LOGIN_GROUP_APPLICATION:
+	 * group=<gid>:app=<application id>
+	 */
+
+	name = kzalloc(TEE_UUID_NS_NAME_SIZE, GFP_KERNEL);
+	if (!name)
+		return -ENOMEM;
+
+	switch (connection_method) {
+	case TEE_IOCTL_LOGIN_USER:
+		scnprintf(name, TEE_UUID_NS_NAME_SIZE, "user=%x",
+			  current_euid().val);
+		break;
+
+	case TEE_IOCTL_LOGIN_GROUP:
+		memcpy(&ns_grp, connection_data, sizeof(gid_t));
+		grp = make_kgid(current_user_ns(), ns_grp);
+		if (!gid_valid(grp)) {
+			rc = -EINVAL;
+			goto out;
+		}
+
+		scnprintf(name, TEE_UUID_NS_NAME_SIZE, "group=%x", grp.val);
+		break;
+
+	case TEE_IOCTL_LOGIN_APPLICATION:
+		application_id = tee_session_get_application_id();
+		scnprintf(name, TEE_UUID_NS_NAME_SIZE, "app=%s",
```

Comment:
This will currently result in the string `app=(null)` or similar. I'm not sure this is the best. How about just hardcode it to "app=" until we know what to do with it?

Reply:
Or return error if application ID is null?

Reply:
Yes, that could be an option. We need to be careful to not break compliance testing though.

Reply:
So far the client id handling in kernel is broken... Eg. I would be very surprised that it worked at all for any compliance testing so I do not see that as a worry.

Reply:
Give that the resulting UUID passed to the TA is not specified the GP tests may very well be satisfied with not returning an error.

Reply:
Probably safest would be to return -EINVAL or so if app id is NULL so that there would not be safe feeling that everything is working correctly.